### PR TITLE
Only perform routehandle calculation once per connection if the field…

### PR DESCRIPTION
**Pull request by feiliuesmf**
The code implemented in this branch provides a mechanism for the user to specify that the fields on src/dst sides of data exchange are created in such a way that all fields on the src side are created on the same grid (or mesh etc) and all fields on the dst side are created on the same grid (or mesh etc). Because the user is knowledgeable of the model set up and can specify this condition correctly instead of relying on grid/mesh comparison within the nuopc connector code.

NRL has run this code for a few months by now and this implementation has trimmed 20-30 minutes per cpu per forecast which adds up significantly to millions of CPU hours saved for the simulations performed at NRL. We are confident that the code has been performing correctly and other users of ESMF/NUOPC can benefit from this feature to have a significant saving on CPU time.

### CODE REVIEW

File: src/addon/NUOPC/src/NUOPC_Connector.F90 [hunk 1]
```diff
@@ -7931,6 +7973,16 @@ subroutine FieldBundleCplStore(srcFB, dstFB, cplList, rh, zeroRegions, &
...
           if (ESMF_LogFoundError(rcToCheck=localrc, msg=ESMF_LOGERR_PASSTHRU, &
             line=__LINE__, file=trim(name)//":"//FILENAME, rcToReturn=rc)) return  ! bail out
+          if (btest(verbosity,12)) then
+            write(msgString,*) "   After FieldRegridStore: "//trim(name)//"_"//trim(chopStringList(1))
+            call ESMF_LogWrite(msgString, ESMF_LOGMSG_INFO)
+          endif
+          rhComputed = .true.
+          rhhSaved = rhh
+         else !
+          rhh = rhhSaved
```
[the remark this answers is not in the document]
Comment (feiliuesmf) [reply src/addon/NUOPC/src/NUOPC_Connector.F90:7983]: Dan, the field pairs are iterated, the logic in this code is similar to
if this is the first field pair, then
   calculate rh
   flag rhComputed to true
   save the rh
else // for subsequent pairs
   use the rh saved

rhhSaved allows all subsequent field pair interpolation to proceed without recomputing rh.

**Comment by theurich** [comment]
@feiliuesmf, I am trying to understand why the internally implemented mechanism for automatic re-use of RHs is not working for the ESPC cases, where the proposed `singleRH` is giving you better performance. Are these cases using Grids, Meshes, LocStream?

**Comment by theurich** [comment]
And I assume that you need regridding, i.e. you are not setting `remapMethod=Redist`?

**Comment by feiliuesmf** [comment]
Regridding for this case.

**Comment by theurich** [comment]
@feiliuesmf, please test out https://github.com/esmf-org/esmf/pull/114 with a few of the relevant ESPC cases. This new PR extends the automatic Regrid-RH-reuse optimization under NUOPC_Connector to cover and Grid/Mesh combinations between source and destination. It should in principle give you the same savings as you have observed with the proposed `singlerh` approach.

The advantage of the automatic approach is that it does not require manually setting a specific flag. It therefore transparent to the user. It is more general: the NUOPC_Connector will automatically use the least number of Store() calls it can, even if some src/dst pairs do need separate RHs, but some pairs can re-use. It is more robust in basing the decision on the actual src/dst pairs encountered, rather a single flag that applies to all pairs. 

Please let us know how https://github.com/esmf-org/esmf/pull/114 tests out for the ESPC cases.

**Comment by feiliuesmf** [comment]
Thanks Gerhard, I didn't follow the github workflow, when you say #114, is
it a tag/branch of ESMF that I could try, is it feature/reuse-RH?

**Comment by theurich** [comment]
@feiliuesmf, yes, that is right, the #114 is a reference to another PR, and the associated branch is https://github.com/esmf-org/esmf/tree/feature/reuse-RH. That's the code to try out. Thanks.

**Comment by feiliuesmf** [comment]
Gerhard, just to be sure, setting big 12 results in the following:
  Verbosity = 1000

I didn't see how that works in the doc but examples seem to show this is
how the binary is coded.

**Comment by theurich** [comment]
@feiliuesmf - see the table at https://earthsystemmodeling.org/docs/nightly/develop/NUOPC_refdoc/node3.html#SECTION00033400000000000000. Verbosity is a string, that is converted to an integer, and then interpreted as a bit field. You can just use one of the string presets, like "high" on the user side, and it will set bit-12 (among some others). Or if you only want to set bit-12, you'd set Verbosity to 2^12, i.e. `Verbosity="4096"`.

**Comment by feiliuesmf** [comment]
I see, I think it'd be clearer if the doc says decimal integer.

**Comment by theurich** [comment]
Fei has confirmed that branch feature/reuse-RH, i.e. PR #114, results in comparable performance as the approach proposed in this PR. Closing this PR and finish under PR #114.

**Comment by theurich** [comment]
Fei has confirmed that branch feature/reuse-RH, i.e. PR #114, results in comparable performance as the approach proposed in this PR. Closing this PR and finish under PR #114.

**Pull request closed without merging**